### DISCUSSION TITLE
Has official support been considered for the Mock Service Worker (MSW) Storybook addon been considered?

### DISCUSSION THREAD

**Discussion by ontoneio** [storybookjs/storybook#26456] [Help] [answered]
### Summary

The official [storybook testing handbook](https://storybook.js.org/tutorials/ui-testing-handbook/react/en/composition-testing/), suggests the use of the [Mock Service Worker (MSW) addon](https://storybook.js.org/addons/msw-storybook-addon/) to use for composition testing.

The library is however not officially released as module of the Storybook monorepo, and therefore it frequently falls out of version alignment. The library still uses an older `1.0` version of `MSW` library where we already have a new major version. The lack of parity between the testing handbook tutorial guidance and the maintenance of the library makes an inconsistent Storybook development experience.

### Additional information

Errors like the following are a constant issue with the Storybook release cycle. Add-ons that give needed functionality to SB are broken due to peer dependency issues and out of date libraries.

```bash
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: msw@1.3.2
npm ERR! Found: typescript@5.3.3
npm ERR! node_modules/typescript
npm ERR!   dev typescript@"5.3.3" from the root project
npm ERR!   typescript@"5.3.3" from @microsoft/api-extractor@7.39.0
npm ERR!   node_modules/@microsoft/api-extractor
npm ERR!     @microsoft/api-extractor@"7.39.0" from vite-plugin-dts@3.7.3
npm ERR!     node_modules/vite-plugin-dts
npm ERR!       dev vite-plugin-dts@"3.7.3" from the root project
npm ERR!   21 more (@nx/eslint, @phenomnomnominal/tsquery, ...)
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peerOptional typescript@">= 4.4.x <= 5.2.x" from msw@1.3.2
npm ERR! libs/civic-ui/node_modules/msw
npm ERR!   peer msw@">=0.35.0 <2.0.0" from msw-storybook-addon@1.10.0
npm ERR!   libs/civic-ui/node_modules/msw-storybook-addon
npm ERR!     dev msw-storybook-addon@"*" from @alcoitd/civic-ui@0.1.0
npm ERR!     libs/civic-ui
npm ERR!       @alcoitd/civic-ui@0.1.0
npm ERR!       node_modules/@alcoitd/civic-ui
npm ERR! 
npm ERR! Conflicting peer dependency: typescript@5.2.2
npm ERR! node_modules/typescript
npm ERR!   peerOptional typescript@">= 4.4.x <= 5.2.x" from msw@1.3.2
npm ERR!   libs/civic-ui/node_modules/msw
npm ERR!     peer msw@">=0.35.0 <2.0.0" from msw-storybook-addon@1.10.0
npm ERR!     libs/civic-ui/node_modules/msw-storybook-addon
npm ERR!       dev msw-storybook-addon@"*" from @alcoitd/civic-ui@0.1.0
npm ERR!       libs/civic-ui
npm ERR!         @alcoitd/civic-ui@0.1.0
npm ERR!         node_modules/@alcoitd/civic-ui
```

### Create a reproduction

In the repro link below run CLI command `npm i msw msw-storybook-addon -D` and read the error.

https://stackblitz.com/edit/github-hgybva?file=package.json&preset=node

**Comment by valentinpalkovic** [comment, marked as answer]
Thanks a ton for bringing this up! It’s super helpful to talk about the ups and downs of using the Mock Service Worker (MSW) with Storybook. I get that it’s been a bit bumpy, especially when the guidance from the Storybook team doesn’t quite match up with how things are running. It’s frustrating when you’re trying to keep everything updated.

We like to test new stuff as separate addons before deciding if they should become a permanent part of the Storybook family. This way, we can see if you and other developers find them useful without overwhelming everyone with changes. It’s all about finding that sweet spot where we add value without making things too complicated and, in the end, committing to something that we can maintain on a longer perspective.

We’re right there with you on needing a more up-to-date version of MSW. There's some cool experimental support for MSW 2.0 going on. You can dive into it with the details found in this PR: https://github.com/mswjs/msw-storybook-addon/pull/122. If you’re up for it, we’d love for you to try it out and let us know what you think. Your feedback could really help!

I totally agree that MSW is key for testing full pages and workflows in Storybook. As we aim to grow from focusing just on components to handling full apps, it makes a lot of sense to have MSW and other important tools right there in our main setup. This way, we can ensure everything works smoothly together with each new Storybook release.

@yannbf Might have a bit more insights of why the addon isn't moved so far into the mono repo and what is actually necessary to move it into it.

**Reply by yannbf** [reply]
Hey @ontoneio thanks for explaining your use case, but please do some research before stating that there is a lack of maintenance. 

The MSW addon is part of the MSW organization. It does not have any dependency on Storybook, therefore there would be no difference if the addon was part of the monorepo or not in regards to the issue at hand. Ever since the release of MSW 2.0 there is a beta version of the addon which supports it. I agree that the documentation could be improved so you would have known about this beta version. There are some challenges regarding portable stories which didn't allow the addon to have a latest release, but that will be hopefully handled soon. You can follow the PR that Valentin linked for updates, and keep using the prerelease version for the time being, which should unblock you.

**Reply by jonniebigodes** [reply]
@ontoneio thank you for reaching out to us with your question. We appreciate it 🙏 ! To give you some context, I've been keeping a close eye on the MSW addon and seeing how we could adjust the documentation accordingly. I refrained from documenting and using the updated MSW version from the tutorials, as there's a bit of work to be done on that front. I didn't want to introduce misconceptions by suggesting pre-release versions. I can assure you that once the updated addon version is released, I'll be more than glad to update the documentation and, in this case, the templates we currently use for the tutorials.

Hope you have a great day.

Stay safe

**Reply by ontoneio** [reply]
Thanks for all the clarifying responses. All the information is very much appreciated.

@yannbf Thanks for the information, but I did in fact research and still was not able to locate the updated beta nor any references to it. When I said lack of maintenance, I meant that there is an obvious disconnect between the MSW versions and that the add-on hadn't been updated in 5 months when there were also major version updates to storybook as well. I am just a regular working developer and don't always have the time to dig deep and find that there is in fact something I can pull in that is semi-stable enough to get me by, I try and spare myself the tech debt personally. 

I am in tons of communities already and sometime information overload parsing through it all can become a drain on ones energy. I didn't mean to suggest that no one was maintaining the package at all. I fully understand the idiosyncrasies of choosing when it is a proper time and coordinating the effort to do so. It's two separate open organizations  making a fantastic thing. I was just not sure where else to ask about this specific integration.

Thanks to all who provided the clarifying information! Super helpful.

**Reply by yannbf** [reply]
Thanks for elaborating! We're all trying our best, it's just the nature of open source, where many projects are being maintained at the same time by not a lot of people. I'll update the README of the MSW addon to communicate about the beta version.
